vault/identity_store_oidc.go

```diff
@@ -0,0 +1,1344 @@
+		Data: map[string]interface{}{
+			logical.HTTPStatusCode:  200,
+			logical.HTTPRawBody:     data,
+			logical.HTTPContentType: "application/json",
+		},
+	}
+
+	return resp, nil
+}
+
+func (i *IdentityStore) pathOIDCIntrospect(ctx context.Context, req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	var claims jwt.Claims
+
+	// helper for preparing the non-standard introspection response
+	introspectionResp := func(errorMsg string) (*logical.Response, error) {
+		response := map[string]interface{}{
+			"active": true,
+		}
+
+		if errorMsg != "" {
+			response["active"] = false
+			response["error"] = errorMsg
+		}
+
+		data, err := json.Marshal(response)
+		if err != nil {
+			return nil, err
+		}
+
+		resp := &logical.Response{
+			Data: map[string]interface{}{
+				logical.HTTPStatusCode:  200,
+				logical.HTTPRawBody:     data,
+				logical.HTTPContentType: "application/json",
+			},
+		}
+
+		return resp, nil
+	}
+
+	rawIDToken := d.Get("token").(string)
+	clientID := d.Get("client_id").(string)
+
+	// validate basic JWT structure
+	parsedJWT, err := jwt.ParseSigned(rawIDToken)
+	if err != nil {
+		return introspectionResp(fmt.Sprintf("error parsing token: %s", err.Error()))
+	}
+
+	// validate signature
+	jwks, err := i.generatePublicJWKS(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	var valid bool
+	for _, key := range jwks.Keys {
+		if err := parsedJWT.Claims(key, &claims); err == nil {
+			valid = true
+			break
+		}
+	}
+
+	if !valid {
+		return introspectionResp("unable to validate the token signature")
+	}
+
+	// validate claims
+	c, err := i.getOIDCConfig(ctx, req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	expected := jwt.Expected{
+		Issuer: c.effectiveIssuer,
+		Time:   time.Now(),
+	}
+
+	if clientID != "" {
+		expected.Audience = []string{clientID}
+	}
+
+	if claimsErr := claims.Validate(expected); claimsErr != nil {
+		return introspectionResp(fmt.Sprintf("error validating claims: %s", claimsErr.Error()))
+	}
+
+	// validate entity exists and is active
+	entity, err := i.MemDBEntityByID(claims.Subject, true)
+	if err != nil {
+		return nil, err
+	}
+	if entity == nil {
+		return introspectionResp("entity was not found")
+	} else if entity.Disabled {
+		return introspectionResp("entity is disabled")
+	}
+
+	return introspectionResp("")
+}
+
+// namedKey.rotate(overrides) performs a key rotation on a namedKey and returns the
+// verification_ttl that was applied. verification_ttl can be overriden with an
+// overrideVerificationTTL value >= 0
+func (k *namedKey) rotate(ctx context.Context, s logical.Storage, overrideVerificationTTL time.Duration) error {
+	verificationTTL := k.VerificationTTL
+
+	if overrideVerificationTTL >= 0 {
+		verificationTTL = overrideVerificationTTL
+	}
+
+	// generate new key
+	signingKey, err := generateKeys(k.Algorithm)
+	if err != nil {
+		return err
+	}
+	if err := saveOIDCPublicKey(ctx, s, signingKey.Public()); err != nil {
```

Comment:
Something to think about for post beta, but we may want a background process running to clean up any keys that we lose reference to due to the below update failing

Reply:
The `expireOIDCPublicKeys()` func that is invoked during `periodicFunc` will do this. It gets a list of all public keys in storage, makes another list of all keys referenced by all roles, and then deletes any unreferenced public keys.

Reply:
Yeah i just got down to that part of the code, sounds good